service/tablet_allocator.cc

```diff
@@ -1128,6 +1127,123 @@ class load_balancer {
+                };
+
+                auto old_tablets_per_shard = get_avg_tablets_per_shard(table_plan.current_tablet_count);
+                old_avg_tablets_per_shard += old_tablets_per_shard;
+                lblogger.debug("old_avg_tablets_per_shard [dc={}, table={}]: {:.3f}", dc, table, old_tablets_per_shard);
+
+                auto tablets_per_shard = get_avg_tablets_per_shard(table_plan.target_tablet_count);
+                avg_tablets_per_shard += tablets_per_shard;
+                lblogger.debug("new_avg_tablets_per_shard [dc={}, table={}]: {:.3f}", dc, table, tablets_per_shard);
+            }
+
+            if (!new_table) {
+                bool overloaded = old_avg_tablets_per_shard > _tablets_per_shard_goal;
+                lblogger.info("avg_tablets_per_shard[dc={}]: {:.3f}{}", dc, old_avg_tablets_per_shard,
+                    overloaded ? " (overloaded!)" : "");
+            }
+
+            bool overloaded = avg_tablets_per_shard > _tablets_per_shard_goal;
+            lblogger.debug("new_avg_tablets_per_shard[dc={}]: {:.3f}{}", dc, avg_tablets_per_shard,
+                overloaded ? " (overloaded!)" : "");
+
+            if (overloaded) {
+                auto scale = _tablets_per_shard_goal / avg_tablets_per_shard;
+
+                for (auto&& [table, table_plan]: plan.tables) {
+                    auto *rs = rs_by_table[table];
+                    if (!rs) {
+                        continue;
+                    }
+
+                    auto rf = rs->get_replication_factor(dc);
+
+                    // If table has no replicas in this DC, scaling it won't help and is harmful to its distribution
+                    // in other DCs.
+                    if (rf) {
+                        if (table_scaling.contains(table)) {
+                            table_scaling[table] = std::min(table_scaling[table], scale);
+                        } else {
+                            table_scaling[table] = scale;
+                        }
+                    }
+                }
+            }
+        }
+
+        for (auto&& [table, scale] : table_scaling) {
+            auto& table_plan = plan.tables[table];
+            auto new_count = std::max<size_t>(1, table_plan.target_tablet_count * scale);
+            lblogger.debug("Scaling down table {} by a factor of {:.3f}: {} => {}", table, scale, table_plan.target_tablet_count, new_count);
+            table_plan.target_tablet_count = new_count;
+        }
+
+        // Generate:
+        //   table_plan.target_tablet_count_aligned
+        //   table_plan.resize_decision
+
+        for (auto&& [table, table_plan] : plan.tables) {
+            table_plan.target_tablet_count_aligned = 1u << log2ceil(table_plan.target_tablet_count);
+
+            if (table_plan.target_tablet_count_aligned > table_plan.current_tablet_count) {
+                table_plan.resize_decision = locator::resize_decision::split();
```

Comment:
This patch is about scaling down, can this case actually happen at this stage?

Reply:
This computes the resize decision after every adjustments, not just scaling. The block was moved here.